src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -682,16 +691,17 @@ private static Task<TResult> RunAsync<TResult>(Func<object, long, TResult> func,
                 }, key, cancellationToken, TaskContinuationOptions.DenyChildAttach, TaskScheduler.Default);
 
                 // If it's possible the task may end up getting canceled, it won't have a chance to remove itself from
-                // the dictionary if it is canceled, so use a separate continuation to do so.
+                // the dictionary and report AfterResolution() if it is canceled, so use a separate continuation to do so.
                 if (cancellationToken.CanBeCanceled)
                 {
-                    task.ContinueWith((task, key) =>
+                    task.ContinueWith(delegate
```

Comment:
Previously this wasn't allocating a new delegate and a new closure on every call. Now it is, as it's closing over the key and the task and the startingTimestamp. It'll even result in that allocation even if !CanBeCanceled, because state it's closing over is at the top-level method scope.

Reply:
At a minimum you can pass the required state in via the object state parameter. That'll require a tuple allocation/boxing, but the one allocation for that is still better than the two allocations for the closure and delegate. And it'll make it specific to the CanBeCanceled case, whereas with how it's currently written the closure allocation will happen first thing in the method body.